host/client_integration_test.go

```diff
@@ -0,0 +1,372 @@
+	}
+}
+
+func (s *clientIntegrationSuite) buildServiceClient() (workflowserviceclient.Interface, error) {
+	cadenceClientName := "cadence-client"
+	cadenceFrontendService := common.FrontendServiceName
+	hostPort := "127.0.0.1:7104"
+
+	ch, err := tchannel.NewChannelTransport(
+		tchannel.ServiceName(cadenceClientName))
+	if err != nil {
+		s.logger.Fatalf("Failed to create transport channel: %v", err)
+	}
+
+	dispatcher := yarpc.NewDispatcher(yarpc.Config{
+		Name: cadenceClientName,
+		Outbounds: yarpc.Outbounds{
+			cadenceFrontendService: {Unary: ch.NewSingleOutbound(hostPort)},
+		},
+	})
+	if dispatcher == nil {
+		s.logger.Fatal("No RPC dispatcher provided to create a connection to Cadence Service")
+	}
+	if err := dispatcher.Start(); err != nil {
+		s.logger.Fatalf("Failed to create outbound transport channel: %v", err)
+	}
+
+	return workflowserviceclient.New(dispatcher.ClientConfig(cadenceFrontendService)), nil
+}
+
+func (s *clientIntegrationSuite) TearDownSuite() {
+	s.host.Stop()
+	s.host = nil
+	s.worker.Stop()
+	s.TearDownWorkflowStore()
+}
+
+func (s *clientIntegrationSuite) SetupTest() {
+	// Have to define our overridden assertions in the test setup. If we did it earlier, s.T() will return nil
+	s.Assertions = require.New(s.T())
+}
+
+func (s *clientIntegrationSuite) TearDownTest() {
+
+}
+
+func (s *clientIntegrationSuite) setupSuite(enableGlobalDomain bool, isMasterCluster bool) {
+	// Have to define our overridden assertions in the test setup. If we did it earlier, s.T() will return nil
+	s.Assertions = require.New(s.T())
+	options := persistence.TestBaseOptions{}
+	options.ClusterHost = "127.0.0.1"
+	options.DropKeySpace = true
+	options.SchemaDir = ".."
+	options.EnableGlobalDomain = enableGlobalDomain
+	options.IsMasterCluster = isMasterCluster
+	s.SetupWorkflowStoreWithOptions(options, nil)
+
+	s.setupShards()
+
+	// TODO: Use mock messaging client until we support kafka setup onebox to write end-to-end integration test
+	s.mockProducer = &mocks.KafkaProducer{}
+	s.mockMessagingClient = mocks.NewMockMessagingClient(s.mockProducer, nil)
+
+	s.host = NewCadence(s.ClusterMetadata, s.mockMessagingClient, s.MetadataManager, s.ShardMgr, s.HistoryMgr, s.ExecutionMgrFactory, s.TaskMgr,
+		s.VisibilityMgr, testNumberOfHistoryShards, testNumberOfHistoryHosts, s.logger, 0, false)
+	s.host.Start()
+
+	s.engine = s.host.GetFrontendClient()
+	s.domainName = "integration-test-domain"
+	s.domainID = uuid.New()
+
+	s.MetadataManager.CreateDomain(&persistence.CreateDomainRequest{
+		Info: &persistence.DomainInfo{
+			ID:          s.domainID,
+			Name:        s.domainName,
+			Status:      persistence.DomainStatusRegistered,
+			Description: "Test domain for integration test",
+		},
+		Config: &persistence.DomainConfig{
+			Retention:  1,
+			EmitMetric: false,
+		},
+		ReplicationConfig: &persistence.DomainReplicationConfig{},
+	})
+}
+
+// testDataConverter implements encoded.DataConverter using gob
+type testDataConverter struct{}
```

Comment:
could you add 2 counter field to testDataConverter and increase them when ToData() and FromData() is called. We need to assert at the end of the test that they are called expected number of times.

Reply:
This doesn't look like integration test, because from this test level you don't care/know what is the correct call times. Also those counters will change if client change somewhere, which confuse people.  
I added a TestClientDataConverter_Failed case to cover that different data converter is actually used.

Reply:
I think it is valuable to verify that count number, and it is also good to verify that we don't call this encoding/decoding more than needed. For example, if there is a regression happen on client code that do this encoding

Reply:
Ok, I am fine with that. Added.